Add first implementation, featuring tests with BATS!

d3adb5 opened the pull request:
Add the first implementation of this wee program. BATS was picked to write
tests. This is the first time I write tests for something written in POSIX sh,
and while it still feels like a hack, it's actually pretty cool!

Closes #2.

d3adb5 committed:
feat: add unimplemented script & tests

Add the single script and test suite, currently without any
implementation whatsoever.

d3adb5 committed:
chore: add bats submodules & move test file

Move the test file to a 'test' directory and add the components of bats
as submodules in this repository. They'll be used as the tests are
written.

d3adb5 committed:
feat: add usage information 'command' & main impl

Add an implementation for the 'main' function and a pseudo-command that
displays usage information for smarky. The implementation might end up
deviating from the intended usage.

d3adb5 committed:
feat: guarantee index file exists

Guarantee the smarky index file exists even if we're not manipulating
its contents, by creating a table using sqlite3 and setting shell option
'-e' to stop at errors.

d3adb5 committed:
feat: implement create command

Implement the create command and add tests for it. String interpolation
and SQL commands is... well, just asking for SQL injection. But hey,
this isn't some sort of professional piece of software, it's just
something for fun running on the user's computer.

d3adb5 committed:
feat: add & implement select command

Add a select command for retrieving commands (and just the commands).

d3adb5 committed:
feat: implement list command

Basic output replacing newlines with double spaces, for an easier time
selecting from something like fzf.

d3adb5 committed:
feat: implement remove command

Implement the remove command, changing the intended usage to a simple id
lookup, rather than using something like "LIKE", which might delete more
than what was intended and adds unnecessary complexity.

d3adb5 committed:
refactor: flip order of 'create' args

It is best that we can set a description when using the editor to add
command bookmarks. Bookmarks with no description aren't very useful,
after all.

d3adb5 committed:
feat: add github actions ci workflow

Add a workflow for continuous integration that runs the tests with BATS.

d3adb5 committed:
feat: add zsh widget to pick from bookmarks

Add a ZSH ZLE widget to pick commands stored as bookmarks in the smarky
index. Bound to C-j by default, might remove it soon in case someone
prefers to bind it to another key combination.

d3adb5 committed:
refactor: rename select function for shellcheck

Rename the select function so that shellcheck doesn't think it is
somehow Bash's select keyword, used for interactive menus implemented at
the interpreter level.

d3adb5 committed:
feat: truncate list output if in a terminal

Truncate the output of the list command if we're running smarky in a
terminal window instead of piping it into another command.

d3adb5 committed:
feat: run shellcheck during ci workflow

Run ShellCheck during the CI workflow.

d3adb5 committed:
feat: use bat for syntax highlighting if installed

If found in $PATH, bat will be used for syntax highlighting, assuming sh
as the language that will be stored in the smarky index.

d3adb5 committed:
feat: add command to update existing bookmarks

Add a command that updates an existing bookmark with a given ID.

d3adb5 committed:
feat: populate temp file w/current cmd when updating

Populate the temporary file used for editing the current command when it
is being updated. This is skipped entirely if the new command line is
provided through the command invocation.

viniciustrainotti commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,37 @@
+  bats:
+    name: BATS
+    runs-on: ubuntu-latest
+
```
I read best practics github wokflow, but I not finding the something timeout for jobs or steps. Nowdays, I seeing examples the in CI's files, sometimes I found timeouts declaratives, well I bielive that you can to add a timeout for steps, as you think needed. e.g:

```timeout-minutes: 2```

d3adb5 replied:
I'd actually never seen that. Guess I can't say I'm that familiar with GitHub Actions after all. It makes sense, considering GitHub Actions billing is calculated by minutes, and while it won't make a difference right now since this is a public repository, it was set in d64cd4f. Thanks!

viniciustrainotti commented on README.md:
```diff
@@ -4,4 +4,6 @@ A shell bookmarker similar to [marker](https://github.com/pindexis/marker), but
 fuzzy matching.
 
+SQLite is used for the index file. Something better probably exists.
+
```
I really really import describe all steps to use this magnifique tool. As person newbies as me, no disturb with questions: "How to install?", "Which is necessary?" .

d3adb5 replied:
I completely forgot to add instructions on how to use or install the tool there. I have plans to add a PKGBUILD for Arch later down the line, then submit it to the AUR. Instructions on what to do right now were added in 52aab6b.

viniciustrainotti commented on test/smarky.bats:
```diff
@@ -0,0 +1,126 @@
+#!/usr/bin/env bats
+
+setup() {
+  load 'helper/bats-support/load'
+  load 'helper/bats-assert/load'
+
+  export SMARKY_INDEX="${BATS_TMPDIR}/smarky-index.db"
+
+  rootDir="$(cd "$(dirname "$BATS_TEST_FILENAME")" && pwd)/.."
+  PATH="$rootDir:$PATH"
+}
+
+teardown() {
+  rm -f "$SMARKY_INDEX"
+}
```
How to work this block code? I checked to work, but how? This BATS behavior?

d3adb5 replied:
This is indeed BATS behavior. Before each `@test`, BATS will run the `setup` function, and after each test, the `teardown` function will be executed.

viniciustrainotti commented on smarky:
```diff
@@ -0,0 +1,97 @@
+#!/bin/sh
+set -e
+
+indexFile="${SMARKY_INDEX:-${XDG_DATA_HOME:-$HOME/.local/share}/smarky-index.db}"
```
You could explain environment variables how to use in README.md

d3adb5 replied:
Done in f6af55c. Didn't expect anybody to use them myself; it was added because it would be easier for testing purposes.

viniciustrainotti commented on zsh/widgets.zsh:
```diff
@@ -0,0 +1,16 @@
+}
+
+zle -N fzf-smarky-pick-command
+bindkey '^J' fzf-smarky-pick-command
```
You could you explain somewhere in the README, the process in the necessary source in terminal or add line in zshrc and source too for use function.

d3adb5 replied:
Done in 52aab6b. Thanks!

viniciustrainotti commented on smarky:
```diff
@@ -0,0 +1,97 @@
+  commandDescription="$1" commandLine="$2" tmpFile="$(mktemp)"
+
+  if [ "$#" -lt 2 ]; then
+    ${EDITOR:-nano} "$tmpFile"
```
Somewhere, set Info for peoples who don't use the EDITOR variable, in case they find it necessary to set EDITOR var to prefer it.

d3adb5 replied:
Added it in f6af55c. Nice catch.

viniciustrainotti commented on smarky:
```diff
@@ -0,0 +1,97 @@
+    commandLine="$(cat "$tmpFile")"
+  fi
+
+  sqlite3 "$indexFile" "
+    insert into commands (command, description) values (
+      '$(echo "$commandLine" | sed "s/'/''/g")',
+      '$(echo "$commandDescription" | sed "s/'/''/g")'
+    );"
+
+  rm "$tmpFile"
```
I missed of answers for the user of whether it worked or not, it is worth giving this feedback to the person who is using the system to have clarity of operation.

d3adb5 replied:
Hm... Personally I prefer to abide by the Unix way: _no news is good news,_ meaning if you get no output out of something like `smarky create "example" "example"`, it means everything went as it was supposed to.

That said, I can see where the rationale might not sit right with some people and am frankly not opposed to the "logging" practice. Clarity of operation is something to consider for human-computer interaction, after all.

Anyways, added that sort of output to the `create` and `update` commands in 1bd2364 and dfdd57c respectively.

d3adb5 committed:
chore: set timeout for ci workflow jobs

Set a timeout for the jobs in the CI workflow. This way we can control
how many minutes they are allowed to spend, even though as of the time
of this commit, GitHub Actions is completely free for public
repositories.

d3adb5 committed:
docs: add install & usage information to README

Add information on how to install and use smarky to the project's
README.md file.

d3adb5 committed:
docs: add env vars information to README

Add information on what environment variables are used and what for in
smarky to the project's README.md

d3adb5 committed:
feat: let user know when a bookmark is created

Let the user know that a bookmark was created successfully, so they're
not left in the dark.

d3adb5 committed:
feat: crash on failure to update, log success otherwise

Return a non-zero exit status when the user tries to update a bookmark
that doesn't exist, but display their success otherwise.

d3adb5 committed:
refactor: add sh extension to temp file

Add a '.sh' extension to the temporary file created whenever bookmarks
are being created or updated, so that text editors have a hint as to
what syntax the file presents itself in.

d3adb5 committed:
feat: test whether the EDITOR is invoked

Test whether the EDITOR is invoked when updating or creating bookmarks.

viniciustrainotti reviewed (approved):
Alright, go ahead.
